File: frontend/coordinator-android/engine/bin/main/ug/qaat/engine/ChronicAbsentee.kt
package ug.qaat.engine

/**
 * Chronic-absentee detection (spec Feature 3) — pure logic over local attendance.
 * Flags students who missed >= [consecutiveThreshold] consecutive sessions, and/or
 * whose attendance % is below [lowAttendancePct]. Both ⇒ CRITICAL, one ⇒ WARNING.
 * Computed on-device from SQLite; no internet needed.
 */
class ChronicAbsentee(
    private val consecutiveThreshold: Int = 3,
    private val lowAttendancePct: Double = 25.0,
) {
    enum class Status { WARNING, CRITICAL }

    data class Flagged(
        val studentId: String,
        val consecutiveMissed: Int,
        val attendancePct: Double,
        val status: Status,
        val lastAttendedSession: String?, // null = never attended
    )

    /**
     * @param sessionsChrono session ids in chronological order (oldest → newest)
     * @param attendedBy     studentId → set of session ids they attended
     * @param students       all enrolled student ids for the unit
     */
    fun evaluate(
        sessionsChrono: List<String>,
        attendedBy: Map<String, Set<String>>,
        students: List<String>,
    ): List<Flagged> {
        if (sessionsChrono.isEmpty()) return emptyList()
        val total = sessionsChrono.size
        val out = mutableListOf<Flagged>()
        for (s in students) {
            val attended = attendedBy[s] ?: emptySet()
            // Consecutive misses, counting back from the most recent session.
            var consec = 0
            for (sid in sessionsChrono.asReversed()) {
                if (sid in attended) break
                consec++
            }
            val pct = attended.size.toDouble() / total * 100.0
            val lowFlag = pct < lowAttendancePct
            val consecFlag = consec >= consecutiveThreshold
            val status = when {
                lowFlag && consecFlag -> Status.CRITICAL
                lowFlag || consecFlag -> Status.WARNING
                else -> null
            } ?: continue
            val lastAttended = sessionsChrono.lastOrNull { it in attended }
            out.add(Flagged(s, consec, pct, status, lastAttended))
        }
        // Most urgent first: CRITICAL before WARNING, then by consecutive misses desc.
        return out.sortedWith(compareByDescending<Flagged> { it.status == Status.CRITICAL }.thenByDescending { it.consecutiveMissed })
    }
}

File: frontend/coordinator-flutter/lib/engine/chronic_absentee.dart
enum AbsenteeStatus { warning, critical }

class Flagged {
  final String studentId;
  final int consecutiveMissed;
  final double attendancePct;
  final AbsenteeStatus status;
  final String? lastAttendedSession;

  const Flagged(
    this.studentId,
    this.consecutiveMissed,
    this.attendancePct,
    this.status,
    this.lastAttendedSession,
  );
}

class ChronicAbsentee {
  final int consecutiveThreshold;
  final double lowAttendancePct;

  ChronicAbsentee({
    this.consecutiveThreshold = 3,
    this.lowAttendancePct = 25.0,
  });

  List<Flagged> evaluate(
    List<String> sessionsChrono,
    Map<String, Set<String>> attendedBy,
    List<String> students,
  ) {
    if (sessionsChrono.isEmpty) return const [];
    final total = sessionsChrono.length;
    final out = <Flagged>[];
    for (final s in students) {
      final attended = attendedBy[s] ?? const <String>{};
      var consec = 0;
      for (final sid in sessionsChrono.reversed) {
        if (attended.contains(sid)) break;
        consec++;
      }
      final pct = attended.length / total * 100.0;
      final lowFlag = pct < lowAttendancePct;
      final consecFlag = consec >= consecutiveThreshold;
      AbsenteeStatus? status;
      if (lowFlag && consecFlag) {
        status = AbsenteeStatus.critical;
      } else if (lowFlag || consecFlag) {
        status = AbsenteeStatus.warning;
      }
      if (status == null) continue;
      final lastAttended = sessionsChrono.lastWhere((sid) => attended.contains(sid),
          orElse: () => '');
      out.add(Flagged(
        s,
        consec,
        pct,
        status,
        lastAttended.isEmpty ? null : lastAttended,
      ));
    }
    out.sort((a, b) {
      final ac = a.status == AbsenteeStatus.critical ? 1 : 0;
      final bc = b.status == AbsenteeStatus.critical ? 1 : 0;
      final byStatus = bc.compareTo(ac);
      if (byStatus != 0) return byStatus;
      return b.consecutiveMissed.compareTo(a.consecutiveMissed);
    });
    return out;
  }
}